From 2213d9e2ab43ac27c85315a4adf252bf6cb9e5ca Mon Sep 17 00:00:00 2001
From: Sebastian Niciarz <sebastian.niciarz@intel.com>
Date: Mon, 29 Apr 2013 15:44:40 +0300
Subject: [PORT FROM MAIN] Telephony: Integrate support for class "l"

BZ: 71460

[DO NOT UPSTREAM]

With this patch, upon receival of ACTIVATE
proactive command, terminal will send
response with result set to "Command beyond
terminal's capabilities."

Code to activate the specific interface
is not part of this patch.

Change-Id: I402c37cce1c9f7596178da5c599bbd68a9434453
Orig-Change-Id: Ifd1b783c9cd3874eff1d5b59bb22441943391034
Signed-off-by: Sebastian Niciarz <sebastian.niciarz@intel.com>
Signed-off-by: Arun Ravindran <arun.ravindran@intel.com>
---
 .../internal/telephony/cat/AppInterface.java       |    3 +-
 .../android/internal/telephony/cat/CatService.java |    4 ++
 .../internal/telephony/cat/CommandParams.java      |    7 ++++
 .../telephony/cat/CommandParamsFactory.java        |   31 ++++++++++++++++++++
 .../telephony/cat/ComprehensionTlvTag.java         |    3 +-
 5 files changed, 46 insertions(+), 2 deletions(-)

diff --git a/src/java/com/android/internal/telephony/cat/AppInterface.java b/src/java/com/android/internal/telephony/cat/AppInterface.java
index 2daf599..5f63e14 100644
--- a/src/java/com/android/internal/telephony/cat/AppInterface.java
+++ b/src/java/com/android/internal/telephony/cat/AppInterface.java
@@ -84,7 +84,8 @@ public interface AppInterface {
         CLOSE_CHANNEL(0x41),
         RECEIVE_DATA(0x42),
         SEND_DATA(0x43),
-        GET_CHANNEL_STATUS(0x44);
+        GET_CHANNEL_STATUS(0x44),
+        ACTIVATE(0x70);
 
         private int mValue;
 
diff --git a/src/java/com/android/internal/telephony/cat/CatService.java b/src/java/com/android/internal/telephony/cat/CatService.java
index bcb7792..42a2047 100644
--- a/src/java/com/android/internal/telephony/cat/CatService.java
+++ b/src/java/com/android/internal/telephony/cat/CatService.java
@@ -362,6 +362,10 @@ public class CatService extends Handler implements AppInterface {
 
                 sendTerminalResponse(cmdParams.mCmdDet, ResultCode.OK, false, 0, null);
                 break;
+           case ACTIVATE:
+                sendTerminalResponse(cmdParams.mCmdDet,
+                        ResultCode.BEYOND_TERMINAL_CAPABILITY, false, 0, null);
+                return;
             default:
                 CatLog.d(this, "Unsupported command");
                 return;
diff --git a/src/java/com/android/internal/telephony/cat/CommandParams.java b/src/java/com/android/internal/telephony/cat/CommandParams.java
index 7db768c..a82ed3b 100644
--- a/src/java/com/android/internal/telephony/cat/CommandParams.java
+++ b/src/java/com/android/internal/telephony/cat/CommandParams.java
@@ -221,3 +221,10 @@ class EventListParams extends CommandParams {
     }
 }
 
+class ActivateParams extends CommandParams {
+    int target = 0;
+    ActivateParams(CommandDetails cmdDet, int target) {
+        super(cmdDet);
+        this.target = target;
+    }
+}
diff --git a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
index a4dc082..e6b2e26 100644
--- a/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
+++ b/src/java/com/android/internal/telephony/cat/CommandParamsFactory.java
@@ -186,6 +186,9 @@ class CommandParamsFactory extends Handler {
             case SET_UP_EVENT_LIST:
                  cmdPending = processSetUpEventList(cmdDet, ctlvs);
                  break;
+            case ACTIVATE:
+                 cmdPending = processActivate(cmdDet, ctlvs);
+                 break;
             default:
                 // unsupported proactive commands
                 mCmdParams = new CommandParams(cmdDet);
@@ -1002,4 +1005,32 @@ class CommandParamsFactory extends Handler {
         }
         return false;
     }
+
+    /**
+     * Processes ACTIVATE proactive command from the SIM card.
+     *
+     * @param cmdDet Command Details object retrieved.
+     * @param ctlvs List of ComprehensionTlv objects following Command Details
+     *        object and Device Identities object within the proactive command
+     * @return true if the command is processing is pending and additional
+     *         asynchronous processing is required.
+     */
+    private boolean processActivate(CommandDetails cmdDet,
+            List<ComprehensionTlv> ctlvs) throws ResultException {
+        CatLog.d(this, "process Activate");
+
+        ComprehensionTlv ctlv = searchForTag(ComprehensionTlvTag.ACTIVATE_DESCRIPTOR, ctlvs);
+        if (ctlv != null) {
+            try {
+                if (ctlv.getLength() != 1) {
+                    throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+                }
+                mCmdParams = new ActivateParams(cmdDet,
+                        (ctlv.getRawValue())[ctlv.getValueIndex()]);
+            } catch (IndexOutOfBoundsException e) {
+                throw new ResultException(ResultCode.CMD_DATA_NOT_UNDERSTOOD);
+            }
+        }
+        return false;
+    }
 }
diff --git a/src/java/com/android/internal/telephony/cat/ComprehensionTlvTag.java b/src/java/com/android/internal/telephony/cat/ComprehensionTlvTag.java
index 973dbc8..9e31c27 100644
--- a/src/java/com/android/internal/telephony/cat/ComprehensionTlvTag.java
+++ b/src/java/com/android/internal/telephony/cat/ComprehensionTlvTag.java
@@ -46,7 +46,8 @@ public enum ComprehensionTlvTag {
     LANGUAGE(0x2d),
     URL(0x31),
     BROWSER_TERMINATION_CAUSE(0x34),
-    TEXT_ATTRIBUTE(0x50);
+    TEXT_ATTRIBUTE(0x50),
+    ACTIVATE_DESCRIPTOR(0x7b);
 
     private int mValue;
 
-- 
1.7.4.1

